first code (#2)
lightbend/kalix-trial-scala-shoppingcart: README.md

# Prerequisite
Java 11 or later<br>
SBT 1.3.6 or later<br>
Docker 20.10.14 or higher (to run locally)<br>
# Introduction

If you have gone through the `Explore Kalix Virtaully` journey, you are all set to :
- [Run locally in prod-like environment](#run-locally-in-prod-like-environment) <br>
- [Deploy and run on Kalix Platform on Cloud Provider of your choice ](deploy-and-run-on-kalix-platform-on-cloud-provider-of-your-choice)

> However, if you would like to better understand the `Kalix Developer Experience` and how to build applications on Kalix, here is more details on the same-use.

Reviewer: ```suggestion
However, if you would like to better understand the `Kalix Developer Experience` and how to build applications on Kalix, here is more detail on the same use case.
```

# Kalix Trial - eCommerce - Scala
## Designing Kalix Services
### Use case
![Use case](images/ecommerce-design_kalix_services.png)<br>
eCommerce use case is a simple shopping cart example consisting of product stock and shopping cart.
Product stock models stock (simple quantity) of products that are being sold and Shopping Cart models list of products customer intends to buy.<br>

> In this exercise focus is on implementing Product Stock functionalities

Reviewer: ```suggestion
In this exercise the focus is on implementing Product Stock functionalities
```

#### Product Stock
Data model:
- quantity

**Note**: Product stock data model does not hold the productId because the HTTP/REST API is designed around productId itself.
Operations:
- CREATE/READ/UPDATE/DELETE product stock

### Kalix components
Kalix components are building blocks used to abstract functionalities.<br>

> In this use case we are going to use Kalix component called `Entity`.<br>

Reviewer: ```suggestion
In this use case we are going to use a Kalix component called `Entity`.<br>
```

**Data model**
- quantity

**API**
- HTTP/RES
- Endpoints:
    - **Create**
      `POST /product-stock/{productId}/create`

      Request (JSON):
        - quantity (int)

      Response (JSON): {}
    - **Read**
      `GET /product-stock/{productId}/get`
      Request (JSON): No body

      Response (JSON):
        - quantity (int)
    - **Update**
      `PUT /product-stock/{productId}/update`

      Request (JSON):
        - quantity (int)

      Response (JSON): {}
    - **Delete**
      `DELETE /product-stock/{productId}/delete`

      Request (JSON): No body

      Response (JSON): {}

## Kickstart Kalix development project
### Kalix giter8 template
Kalix [giter8](https://www.foundweekends.org/giter8/) generates a new Scala development project from Kalix template
### Create shopping cart Maven project from Kalix template
Execute in command line:
```
sbt new lightbend/kalix-value-entity.g8
```
Use this setup:
```
name [My Kalix Project]: kalix-trial-shoppingcart-scala
sbt_version [1.9.2]: <enter>
scala_version [2.13.10]: <enter>
package [com.example]: com.example.shoppingcart
```
giter8 generates Scala project:

> ![Scala project structure](images/ecommerece-giter8-project-structure.png)

Reviewer: Need a screenshot of the Scala project structure.

- `src/test/scala`
  - `ProductStockEntitySpec` scalatest spec class for unit testing (covered in Test - Unit test section)
  - `ProductStockServiceIntegrationSpec` scalatest spec class for integration testing (covered in Test - Integration test section)

**Note:** Generated code in `src` will not be overwritten/regenerated in next in compile but only codebase in `target`.<br>
### ProductStockEntity business logic
For each `Entity` emptyState method, that returns an empty `ProductStock` needs to be implemented. We can use protobuf generated `defaultInstance`.
```
override def emptyState: ProductStock = ProductStock.defaultInstance
```
#### `create` endpoint
Business logic for create is to persist product stock data if not yet exists. In other cases returns an ERROR.
```
override def create(currentState: ProductStock, createProductStockRequest: CreateProductStockRequest): ValueEntity.Effect[Empty] = 
if (currentState == ProductStock.defaultInstance)
  effects.updateState(ProductStock.of(createProductStockRequest.quantity)).thenReply(Empty.defaultInstance)
else
  effects.error("Already created")
```
### `get` endpoint

> Business logic for get is to product stock data if exists and if not return not found error.

Reviewer: ```suggestion
Business logic for get is to get product stock data if it exists and if not return not found error.
```

  effects.error("Not found", Status.Code.NOT_FOUND)
else
  effects.reply(currentState)

```
### `update` endpoint
Business logic for update is to update product stock data if product was already created. If product is not found, return NOT FOUND error.
```
override def update(currentState: ProductStock, updateProductStockRequest: UpdateProductStockRequest): ValueEntity.Effect[Empty] =
if (currentState == ProductStock.defaultInstance)
  effects.error("Not found", Status.Code.NOT_FOUND)
else
  effects.updateState(currentState.withQuantity(updateProductStockRequest.quantity)).thenReply(Empty.defaultInstance)
```
### `delete` endpoint
Business logic for delete is delete data if product stock exists and return NOT FOUND error if not.
Here the soft delete is done by updating the state to `ProductStock.defaultInstance`.
```
override def delete(currentState: ProductStock, deleteProductStockRequest: DeleteProductStockRequest): ValueEntity.Effect[Empty] =
if (currentState == ProductStock.defaultInstance)
  effects.error("Not found", Status.Code.NOT_FOUND)
else
  effects.updateState(ProductStock.defaultInstance).thenReply(Empty.defaultInstance)
```
## Test

> Kalix comes with very rich test kit for unit and integration testing of Kalix code

Reviewer: ```suggestion
Kalix comes with very rich test kit for unit and integration testing of Kalix code.
```
